File: bwi_tasks/src/scav_hunt.py
#!/usr/bin/python

import rospy
import sys
import subprocess # for openning text and image files
import os # for listing files in a directory

from segbot_gui.srv import *

def scav_hunt():

    rospy.wait_for_service('question_dialog')

    # tasks:
    # 0, blueshirt
    # 1, interaction
    # 2, whiteboard

    tasks = []
    tasks.append('* find a person wearing blue shirt')
    tasks.append('* interact with a person through natural language (testing)')
    tasks.append('* find a person standing in front of a whiteboard')
    tasks.append('* take a picture of something with a "Starbucks Coffee" logo')

    task_names = ['blueshirt', 'interaction', 'whiteboard', 'logo']

    blueshirt_pic = ''
    interaction_txt = ''
    whiteboard_pic = ''
    path_to_files = '/home/bwi/Desktop/'

    while not rospy.is_shutdown():
    
        files = os.listdir(path_to_files)

        finished = 'Finished tasks:'
        todo = '\nTodo tasks:'
        ending = '\nClick the buttons to see how I performed in the tasks\n'
        buttons = []

        for f in files:
            if f.find('blueshirt') >= 0:
                blueshirt_pic = f
                finished = finished + '\n\t' + tasks[0]
                buttons.append('blueshirt')
                break
        else:
            todo = todo + '\n\t' + tasks[0]

        for f in files:
            if f.find('interaction') >= 0:
                interaction_txt = f
                finished = finished + '\n\t' + tasks[1]
                buttons.append('interaction')
                break
        else:
            todo = todo + '\n\t' + tasks[1]

        for f in files:
            if f.find('whiteboard') >= 0:
                whiteboard_pic = f
                finished = finished + '\n\t' + tasks[2]
                buttons.append('whiteboard')
                break
        else:
            todo = todo + '\n\t' + tasks[2]

        for f in files:
            if f.find('logo') >= 0:
                logo_pic = f
                finished = finished + '\n\t' + tasks[3]
                buttons.append('logo')
                break
        else:
            todo = todo + '\n\t' + tasks[3]

        # print to screen
        try: 
            handle = rospy.ServiceProxy('question_dialog', QuestionDialog)

            # in case some tasks have been finished
            if finished != finished_last: 

                if len(finished) > 20: 
                    res = handle(1, finished + '\n' + todo + '\n' + ending, buttons, 5)
                else:
                    res = handle(1, finished + '\n' + todo, buttons, 5)
    
        except rospy.ServiceException, e:
            ROS_ERRR("Service call failed: %s"%e)

        # used for checking if necessary to update segbot_gui
        finished_last = finished

        # which picture/text to view? 
        if res.index < 0: 
            continue
        elif buttons[res.index] == 'blueshirt':
            img = subprocess.Popen(["eog", path_to_files + blueshirt_pic])
        elif buttons[res.index] == 'interaction':
            img = subprocess.Popen(["gedit", path_to_files + interaction_txt])
        elif buttons[res.index] == 'whiteboard':
            img = subprocess.Popen(["eog", path_to_files + whiteboard_pic])
        elif buttons[res.index] == 'logo':
            img = subprocess.Popen(["eog", path_to_files + logo_pic])

if __name__ == "__main__":

    scav_hunt()

File: bwi_tasks/src/scav_hunt.cpp

#include "ros/ros.h"
#include "std_msgs/String.h"
#include "segbot_gui/QuestionDialog.h"

#include <string>
#include <stdlib.h>
#include <vector>

std::string status_shirt, status_object, status_board;
std::string shirt_color, object_name;
std::string file_shirt, file_object, file_board; 


void callback_shirt(const std_msgs::String::ConstPtr& msg)
{
 
    if (msg->data.find("running") != std::string::npos) 
        status_shirt = "ongoing";
    else {
        status_shirt = "finished";
        int pos = msg->data.find(":"); 
        file_shirt = msg->data.substr(pos + 1); 
    }
        
}

void callback_object(const std_msgs::String::ConstPtr& msg)
{
 
    if (msg->data.find("running") != std::string::npos)       
        status_object = "ongoing";
    else {
        status_object = "finished";
        int pos = msg->data.find(":");
        file_object = msg->data.substr(pos + 1);
    }

}

void callback_board(const std_msgs::String::ConstPtr& msg)
{
 
    if (msg->data.find("running") != std::string::npos)       
        status_board = "ongoing";
    else {
        status_board = "finished";
        int pos = msg->data.find(":");
        file_board = msg->data.substr(pos + 1);
    }
        
}

int main(int argc, char **argv){

    ros::init(argc, argv, "scav_hunt");
    ros::NodeHandle nh;

    ros::ServiceClient client = 
            nh.serviceClient<segbot_gui::QuestionDialog>("question_dialog");
    segbot_gui::QuestionDialog srv; 

    ros::Subscriber sub1 = nh.subscribe("segbot_blue_shirt_status", 1000,
                                        callback_shirt);

    ros::Subscriber sub2 = nh.subscribe("segbot_object_detection_status", 1000,
                                        callback_object);
 
    ros::Subscriber sub3 = nh.subscribe("segbot_whiteboard_status", 1000,
                                        callback_board);
        
    ros::Rate rate(10); 

    std::string message_finished, message_todo, message_ending; 
    std::vector<std::string> buttons; 

    message_finished = "Finished tasks:";
    message_todo = "\nTodo tasks:";
    message_ending = "\nClick buttons to see how I performed in the tasks\n";

    while (ros::ok())
    {
        
        rate.sleep();       
        ros::spinOnce();

        if (status_shirt.find("ongoing") != std::string::npos) 
            message_todo += "\n\t* find a person wearing '" +
                            shirt_color + " shirt"; 
        else 
        {
            message_finished += "\n\t* find a person wearing '" + 
                                shirt_color + " shirt"; 
            buttons.push_back("color shirt");
        }
 
            
        if (status_object.find("ongoing") != std::string::npos) 
            message_todo += "\n\t* take a picture of object '" + object_name; 
        else 
        {
            message_finished += "\n\t* take a picture of object '" + object_name;
            buttons.push_back("object detection");
        }


        if (status_board.find("ongoing") != std::string::npos) 
            message_todo += "\n\t* find a person standing near a whiteboard"; 
        else 
        {
            message_finished += "\n\t* find a person standing near a whiteboard"; 
            buttons.push_back("whiteboard");
        }


        if (message_finished.length() + message_todo.length() < 30) {
            
            srv.request.type = 0;
            srv.request.message = message_finished + "\n" + message_todo + 
                                  "\n" + message_ending; 
        }
        else 
        {
            srv.request.type = 1;
            srv.request.message = message_finished + "\n" + message_todo; 
            srv.request.options = buttons; 
        }

        if (!client.call(srv)) 
        {
            ROS_ERROR("Failed to call service question_dialog"); 
            return 1;
        }

        std::string eog = "eog ";

        if (srv.response.index < 0) 
            continue;

        else if (buttons[srv.response.index].find("shirt") != std::string::npos)
            system((eog + file_shirt).c_str());

        else if (buttons[srv.response.index].find("object") != std::string::npos)
            system((eog + file_object).c_str());

        else if (buttons[srv.response.index].find("board") != std::string::npos)
            system((eog + file_board).c_str());

    }
}





